### PR TITLE
cmake: disable_all_modules support

### PR DESCRIPTION
**Pull request by furiel**
- If `-DDISABLE_ALL_MODULES=yes` specified for cmake, all switchable modules are disabled by default. One can still override specific modules by: `-DDISABLE_ALL_MODULES=yes -DENABLE_PYTHON=yes`. The order of definitions do not matter.
- Fixed a bug, when `-DENABLE_IPV6=no` resulted in compile error in a unit test.
- Fixed a bug when json unit tests were still tried to compiled even without json support.
- Removed `-DENABLE_BASICFUNCS`, as it causes compile error in some unit tests, it is not switchable in autotools, and does not have any external dependencies.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: cmake/module_switch.cmake [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+# COPYING for details.
+#
+#############################################################################
+
+function (module_switch NAME HELP_STRING)
+  if (DEFINED ${NAME})
+    return()
+  endif()
+
+  if (DISABLE_ALL_MODULES)
```
Comment (Kokan) [inline cmake/module_switch.cmake]: This way the `DISABLE_ALL_MODULES` is not visible on gui/tui (like `ccmake`). I would prefer to have all user option to be visible. What do you think ?

Comment (furiel) [reply cmake/module_switch.cmake]: Agreed. I will do so.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Pull request merged**